pkg/guestagent/kubernetesservice/kubernetesservice.go

```diff
@@ -0,0 +1,152 @@
+package kubernetesservice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type Protocol string
+
+const (
+	// UDP/SCTP when lima port forwarding works on those protocols
+	TCP Protocol = "TCP"
+)
+
+type Entry struct {
+	Protocol Protocol
+	IP       net.IP
+	Port     uint16
+}
+
+type ServiceWatcher struct {
+	rwMutex         sync.RWMutex
+	serviceInformer cache.SharedIndexInformer
+}
+
+func NewServiceWatcher() *ServiceWatcher {
+	return &ServiceWatcher{}
+}
+
+func (s *ServiceWatcher) setServiceInformer(serviceInformer cache.SharedIndexInformer) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.serviceInformer = serviceInformer
+}
+
+func (s *ServiceWatcher) getServiceInformer() cache.SharedIndexInformer {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.serviceInformer
+}
+
+func (s *ServiceWatcher) Start() {
+	retryInterval := 10 * time.Second
+	wait.PollInfinite(retryInterval, func() (done bool, err error) {
+		kubeClient, err := tryGetKubeClient()
+		if err != nil {
+			logrus.Infof("failed to get kube client: %v, will retry in %v", err, retryInterval)
+			return false, nil
+		}
+
+		ctx := context.TODO()
+		informerFactory := informers.NewSharedInformerFactory(kubeClient, time.Hour)
+		serviceInformer := informerFactory.Core().V1().Services().Informer()
+		informerFactory.Start(ctx.Done())
+		cache.WaitForCacheSync(ctx.Done(), serviceInformer.HasSynced)
+
+		s.setServiceInformer(serviceInformer)
+		return true, nil
+	})
+}
+
+func tryGetKubeClient() (kubernetes.Interface, error) {
+	candidateKubeConfigs := []string{
+		"/etc/rancher/k3s/k3s.yaml",
+		"/root/.kube/config",
+	}
+
+	for _, kubeconfig := range candidateKubeConfigs {
+		_, err := os.Stat(kubeconfig)
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			return nil, fmt.Errorf("stat kubeconfig %s failed: %w", kubeconfig, err)
+		}
+
+		restConfig, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			return nil, fmt.Errorf("build kubeconfig from %s failed: %w", kubeconfig, err)
+		}
+		u, err := url.Parse(restConfig.Host)
+		if err != nil {
+			return nil, fmt.Errorf("parse kubeconfig host %s failed: %w", restConfig.Host, err)
+		}
+		if u.Hostname() != "127.0.0.1" { // might need to support IPv6
+			// ensures the kubeconfig points to local k8s
+			continue
+		}
+
```

Comment:
What you have here seems fine, but, want to let you know that you can also leverage the library to do more work for you, something like:
```
func getClientConfig(configPath string) (*restclient.Config, error) {
	loadingRules := clientcmd.ClientConfigLoadingRules{
		ExplicitPath: configPath,
	}
	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&loadingRules, nil)
	config, err := clientConfig.ClientConfig()
	if err != nil {
		return nil, fmt.Errorf("could not load Kubernetes client config from %s: %w", configPath, err)
	}

	return config, nil
}
```

Reply:
Nice one, I wasn't aware of this approach. Will update the code later today.

Reply:
Hmmm, after some trying I would like to stay at current implementation. Because

using `Precedence` makes error message kind of confusing
```go
	loadingRules := clientcmd.ClientConfigLoadingRules{
		Precedence: []string{
			"/etc/rancher/k3s/k3s.yaml",
			"/root/.kube/config",
		},
	}
```

using `ExplicitPath` needs the same amount of code, and it would rely on client-go to return os.ErrNotExist if `configPath` is missing.
```go
	loadingRules := clientcmd.ClientConfigLoadingRules{
		ExplicitPath: configPath,
	}
```